### ISSUE TITLE
Migrate remaining tests to `pytest`

### ISSUE THREAD

**Issue by Taragolis** [apache/airflow#29305]
### Body

All tests for Apache Airflow are run using [`pytest`](http://doc.pytest.org/en/latest/). We have about 80 tests modules which still use [`unittests`](https://docs.python.org/3/library/unittest.html) runner for pytest , initial plan was "convert all unit tests to standard "asserts" semi-automatically, but this will be done later in Airflow 2.0 development phase. That will include setUp/tearDown/context managers and decorators.", see: [Writing Unit Tests](https://github.com/apache/airflow/blob/main/TESTING.rst#id4)

Personally I do not have a free time to complete this tasks, if someone would like to work on one (or many) parts of this Task just let us know in comments or _create a PR with reference to this Issue_.

For keep in track remaining tasks for migration to "native" pytest runner I split it by separate phases

Phase 1: Migration unittests to pytest
---

1. Replace unittests by pytest
2. Replace `@parameterized.expand` by `@pytest.mark.parametrize`

Example how it done previously: https://github.com/apache/airflow/pulls?q=is%3Apr+migrate+tests+pytest+is%3Aclosed+

**Always**
- [x] Migrate remaining `tests/always` to pytest #29375

**CLI**
- [x] Migrate remaining `tests/cli` to pytest #29354

**Other**
- [x] Migrate remaining `tests/operators` to pytest https://github.com/apache/airflow/pull/29377
- [x] Replace `@parameterized.expand` by `@pytest.mark.parametrize` in `tests/callbacks` https://github.com/apache/airflow/pull/29373

**Providers**
- Google (GCP)
   + [x] Migrate remaining GCP operators tests `tests/providers/google/cloud/operators` to pytest #29775
   + [x] Migrate remaining GCP transfers operators tests `tests/providers/google/cloud/transfers` to pytest #29757
   + [x] Migrate remaining GCP sensors tests `tests/providers/google/cloud/sensors` to pytest #29642
- Amazon (AWS)
   +  [x] Migrate remaining AWS tests `tests/providers/amazon/aws` to pytest #29758
- Apache Flink
  + [x]  Migrate Apache Flink tests `tests/providers/apache/flink` to pytest #29604

Phase 2: Ban unittests runner
---

It should be done after Phase 1.

This also include remove `parametrized` package, all features already exists in pytest core, see: [How to parametrize fixtures and test functions](https://docs.pytest.org/en/latest/how-to/parametrize.html#how-to-parametrize-fixtures-and-test-functions)

- [x] Remove all unittests.TestCase usage in `tests/test_utils`, as far as I know right now [only one place](https://github.com/Taragolis/airflow/blob/ce677862be4a7de777208ba9ba9e62bcd0415393/tests/test_utils/asserts.py#L38) where we use it.  https://github.com/apache/airflow/pull/30685#pullrequestreview-1388630869
- [x] Ban inheritance from `unittest.TestCase` it could be done by different ways:
   + Create collector hook which raise an error if detected unittests.TestCase, the opposite how builtin pytest plugin works, see: [unittest.py::pytest_pycollect_makeitem](https://github.com/pytest-dev/pytest/blob/4a46ee8bc957b06265c016cc837862447dde79d2/src/_pytest/unittest.py#L44-L57)
   + Develop pre-commit hook https://github.com/apache/airflow/pull/30775
- [ ] Disable builtin unittests plugin by add `-p no:unittest` in [`pytest.ini` adopts section](https://github.com/apache/airflow/blob/d80b583db07197c8c3d0549a805e83ceaaf10d52/pytest.ini#L19-L26), this step only disable discovering classes based on `unittests.TestCase`.

Some useful stuff
---

I use this command to find any remaining usage of `unittests.TestCase` and `parametrized`

```console

# All modules which potentially use  unittests.TestCase
❯ grep -rl 'TestCase' ./tests | sort | uniq -c | sort -nr

# All modules which potentially use parameterized
❯ grep -rl 'from parameterized' ./tests | sort | uniq -c | sort -nr

# List of providers which potentially use  unittests.TestCase 
❯ grep -rl 'TestCase' tests/providers/ | cut -d"/" -f4 | sort | uniq -c | sort -nr

```

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by Abhishek-kumar-samsung** [comment]
I would like to work on this issue. I am checking the code for this.

**Comment by Abhishek-kumar-samsung** [comment]
I have migrated one file from unittest to pytest.

PR => [https://github.com/apache/airflow/pull/29354]

**Comment by Abhishek-kumar-samsung** [comment]
@Taragolis did changes as you asked, regarding parameterized.expand in test cli files.

**Comment by Taragolis** [comment]
@Abhishek-kumar-samsung You could comment in PR, you really don't need to ping anywhere 🤣 
I've have a full time job (as many other contributors of Airflow) so this is regular thing if you do not have feedback quick, but sooner or later I or someone else look on PR and related issues 👍

**Comment by Abhishek-kumar-samsung** [comment]
Oh sorey @Taragolis this is my first time with open source thats why i am not more comfortable with procedure, will take care next time.

**Comment by Abhishek-kumar-samsung** [comment]
@Taragolis tests in always are also completely migrated to pytest as per me. Only mock objects of unittest are still used, other things migrated.

So i think you can tick mark that task also.

**Comment by Taragolis** [comment]
Thanks for contribution and help with migrations 💯 🥇 

> Only mock objects of unittest are still used, other things migrated.

That is fine we do not have plan to remove them, even [`pytest-mock`](https://pypi.org/project/pytest-mock/) use internally `unittests.mock`

**Comment by JammUtkarsh** [comment]
Hey @Taragolis,
If I were to contribute to this issue, would I require the access to cloud environment like GCP or AWS?
ref: Phase 1/Provider

**Comment by Taragolis** [comment]
Unit tests not required access to cloud environments.
For example configuration of Amazon Provider unit tests disable access to real AWS environment.

**Comment by Abhishek-kumar-samsung** [comment]
@Taragolis regarding 

`Remove all unittests.TestCase usage in tests/test_utils, as far as I know right now [only one place](https://github.com/Taragolis/airflow/blob/ce677862be4a7de777208ba9ba9e62bcd0415393/tests/test_utils/asserts.py#L38) where we use it.`

Is test_utils also any testcase file?

because when i ran it , it runs zero test cases.

**Comment by auvipy** [comment]
is this issue complete to be closed or need some more work?

**Comment by Taragolis** [comment]
Also nice to have solve Phase 2

**Comment by auvipy** [comment]
so I can handle the phase two

**Comment by eladkal** [comment]
@auvipy are you raising PR for phase 2?

**Comment by vincbeck** [comment]
You can mark phase 2.2 as done ;)

**Comment by eladkal** [comment]
> Disable builtin unittests plugin by add -p no:unittest in [pytest.ini adopts section](https://github.com/apache/airflow/blob/d80b583db07197c8c3d0549a805e83ceaaf10d52/pytest.ini#L19-L26), this step only disable discovering classes based on unittests.TestCase.

About this step the file was removed in https://github.com/apache/airflow/pull/30315


--------

thank you @vincbeck
